### ISSUE TITLE
Keeping history clean

### ISSUE THREAD

**Issue by dariusf** [HubTurbo/HubTurbo#806]
In the spirit of keeping history tidy, could I suggest a couple of things we can try?

Here's an example of a not-so-good commit: 582dd80a33103d1ae20d2e939011adf52fab6fb8. It describes what was done but at the wrong level of abstraction.

Commits are meant to record history for the benefit of maintainers. Someone digging through history likely won't be interested in a local, obvious change like this. That suggests that this change should have been part of a larger one.

So:

- **Squash noisy commits** on feature branches before they are merged. These are things like attempts to appease Travis, linting errors, responses to reviews, formatting changes, etc. Basically, rewrite history within the branch to remove anything irrelevant to a future reader. *Splitting* commits would be ideal but you don't have to do it.
- **Rebase onto master**, instead of merging master into branches. How work progressed with respect to the master branch at the time is not very important.

In short, try to make commits as meaningful as possible.

**Comment by damithc** [comment]
I agree. We can add that to the workflow.

**Comment by damithc** [comment]
I considered this sometime back but didn't go ahead. Now I recalled why. While squashing and rebasing is good, I didn't think it was critical. Reason: the branch merge commit gives us a clean view of what happened in that branch. As long as we limit each PR to one issue, when going through the history we just need to see the merge commit and ignore all the other commits in the branch. That's why we require the merge commit message to be in a specific format so that it is easy to find. However, there are places where squashing can be useful. e.g. squashing commits done in response to a review will help reviewer see exactly what changed after the previous review. Going forward, we can try to adopt squashing and rebasing a bit more and see if the extra effort is worth it.

**Comment by ndt93** [comment]
There is a quite popular reading on this topic: [A Git Horror Story](http://mikegerwitz.com/papers/git-horror-story), section on *Managing Large Merges*. I have never seen those more advanced techniques in action through.

**Comment by dariusf** [comment]
Yes, squashing/rebasing in such a manner only tidies up history *within* branches. If the merge commit is large, it may be helpful to see the reasoning behind individual changes. That's the primary motivation.

Squashing everything into one commit neuters `bisect`, as mentioned in the article. The merge commit also works to give us the gist of what was done in a branch.

**Comment by damithc** [comment]
Yes, I think we can definitely be more disciplined in how we create
commits. Let's push in that direction gradually. I'll add some instructions
to the dev guide. We can encourage devs to do,
* Write better commit messages
* Commit at logical points
* Squash if necessary to make the history cleaner
* Squash post-review commits for ease of reviewing
* Encouraged to rebase? not sure about this because rebase vs merge debate
doesn't have a clear winner.

**Comment by dariusf** [comment]
My (subjective) take on the last point when integrating upstream changes is to prefer rebasing, but merge when necessary.

The main difference is that rebasing removes traces of *how* and *what* upstream changes were integrated, while merges preserve them.

Most of the time this history doesn't matter (unrelated upstream changes, or easily-resolved conflicts) and in those cases we can rebase without penalty. Otherwise, if a messy merge conflict could occur, we should merge to document *how* exactly the merge was carried out, so we can revisit it if necessary.
